ENH: (``linalg``) add weighted Gram matrix computation

MengAiDev opened the pull request:
ENH: Add weighted_gram_matrix function to numpy.linalg

Implement weighted_gram_matrix function in numpy/linalg/_linalg.py to
compute weighted Gram matrices efficiently. Add corresponding test cases
in numpy/linalg/tests/test_linalg.py to verify correctness. Update
.gitignore to exclude .env/ directory from version control.

See #29559

MengAiDev committed:
feat(linalg): add weighted Gram matrix computation

- Implement weighted_gram_matrix function in numpy/linalg/_linalg.py
- Add corresponding test cases in numpy/linalg/tests/test_linalg.py
- Update .gitignore to exclude .env/ directory

MengAiDev committed:
lint fix

MengAiDev committed:
fix doc

mattip commented on numpy/linalg/_linalg.py:
```diff
@@ -3659,3 +3659,73 @@ def vecdot(x1, x2, /, *, axis=-1):
+    X = asanyarray(X)
+
+    if X.ndim != 2:
+        raise ValueError("X must be a 2-D array")
```
This could be generalized to higher dimensions, right?

mattip reviewed (commented):
Nice, especially the tests of many of the corner cases. I am not sure we are ready for a Pr to implement this, as the issue has two one-line implementations in the comments, and a fast cython-based implementation is in [tabmat.sandwich](https://tabmat.readthedocs.io/en/latest/api.html#tabmat.MatrixBase.sandwich) with specializations for sparse and categorical matrices. Before this gets much further, I think it would be nice to see some benchmarks of this approach vs. tabmat for some larger matrices.

MengAiDev committed:
feat(linalg): support higher-dimensional arrays in weighted_gram_matrix

- Extend weighted_gram_matrix to handle arrays with more than 2 dimensions
- Apply computation to the last two axes for higher-dimensional arrays
- Update function signature and docstring to reflect new behavior
- Add tests for 3D and 4D arrays
- Improve error handling for mismatched weights shape

MengAiDev committed:
perf(linalg): add performance tests for weighted_gram_matrix

- Add performance tests for medium (1000x500) and large (2000x500) matrices
- Include assertion of result shape and symmetry in performance tests
- Update function docstring to suggest optimized libraries for large matrices

MengAiDev commented:
I update it because when handle small matrixs, numpy is faster, otherwise, tabmat is faster

jorenham commented on .gitignore:
```diff
@@ -132,4 +132,4 @@ tools/swig/test/Vector.py
 tools/swig/test/Vector_wrap.cxx
 tools/swig/test/Array.py
 .openblas
-numpy/_distributor_init_local.py
```
revert please

jorenham reviewed (changes requested):
Like has been said before, this PR is a bit premature, as this should be discussed in the mailing list first. 
But if we assume that there'll be a consensus in favor of this, then  this would also require the stubs to be updated. Specifically in https://github.com/numpy/numpy/blob/8b89d0f127eb76c8f84c14b4814a18814f3bdec7/numpy/linalg/linalg.pyi and https://github.com/numpy/numpy/blob/main/numpy/linalg/_linalg.pyi

MengAiDev commented:
It seems the gitignore is the same as this commit: https://github.com/numpy/numpy/commit/0c203ed4d6ccfb44efef41e87fe5819dbeb7d822

jorenham commented:
> It seems the gitignore is the same as this commit: [0c203ed](https://github.com/numpy/numpy/commit/0c203ed4d6ccfb44efef41e87fe5819dbeb7d822)

well, almost; you're missing a trailing newline

MengAiDev committed:
feat(linalg): add weighted_gram_matrix function

- Implement weighted_gram_matrix in numpy/linalg/_linalg.py
- Add corresponding type annotations in numpy/linalg/_linalg.pyi
- Update numpy/linalg/__init__.py and numpy/linalg/linalg.pyi to include new function
- Improve documentation and examples for weighted_gram_matrix

MengAiDev committed:
Update .gitignore

MengAiDev committed:
Update numpy/linalg/__init__.pyi

Co-authored-by: Joren Hammudoglu <jhammudoglu@gmail.com>

MengAiDev committed:
Update numpy/linalg/_linalg.pyi

Co-authored-by: Joren Hammudoglu <jhammudoglu@gmail.com>

MengAiDev committed:
Update numpy/linalg/_linalg.py

Co-authored-by: Joren Hammudoglu <jhammudoglu@gmail.com>

MengAiDev committed:
Update numpy/linalg/_linalg.py

Co-authored-by: Joren Hammudoglu <jhammudoglu@gmail.com>

MengAiDev committed:
update code

MengAiDev committed:
lint fix
